provider/src/main/kotlin/org/gradle/kotlin/dsl/codegen/ApiTypeProvider.kt

```diff
@@ -0,0 +1,448 @@
+        returnType
+}
+
+
+private
+class TypeSignatureVisitor : SignatureVisitor(ASM6) {
+
+    var isArray = false
+
+    lateinit var binaryName: String
+
+    val typeArguments = ArrayList<TypeSignatureVisitor>(1)
+
+    private
+    var expectTypeArgument = false
+
+    override fun visitBaseType(descriptor: Char) =
+        visitBinaryName(binaryNameOfBaseType(descriptor))
+
+    override fun visitArrayType(): SignatureVisitor =
+        TypeSignatureVisitor().also {
+            visitBinaryName("kotlin.Array")
+            isArray = true
+            typeArguments.add(it)
+        }
+
+    override fun visitClassType(internalName: String) =
+        visitBinaryName(binaryNameOfInternalName(internalName))
+
+    override fun visitInnerClassType(localName: String) {
+        binaryName += "${'$'}$localName"
+    }
+
+    override fun visitTypeArgument() {
+        typeArguments.add(TypeSignatureVisitor().also { it.binaryName = "?" })
+    }
+
+    override fun visitTypeArgument(wildcard: Char): SignatureVisitor =
+        TypeSignatureVisitor().also {
+            expectTypeArgument = true
+            typeArguments.add(it)
+        }
+
+    override fun visitTypeVariable(internalName: String) {
+        visitBinaryName(binaryNameOfInternalName(internalName))
+    }
+
+    private
+    fun visitBinaryName(binaryName: String) {
+        if (expectTypeArgument) {
+            TypeSignatureVisitor().let {
+                typeArguments.add(it)
+                SignatureReader(binaryName).accept(it)
+            }
+            expectTypeArgument = false
+        } else {
+            this.binaryName = binaryName
+        }
+    }
+}
+
+
+private
+fun binaryNameOfBaseType(descriptor: Char) =
+    Type.getType(descriptor.toString()).className
+
+
+private
+fun binaryNameOfInternalName(internalName: String): String =
+    Type.getObjectType(internalName).className
+
+
+private
+fun sourceNameOfBinaryName(binaryName: String): String =
+    when (binaryName) {
+        "void" -> "Unit"
+        "?" -> "*"
+        in collectionTypeStrings.keys -> collectionTypeStrings[binaryName]!!
+        in primitiveTypeStrings.keys -> primitiveTypeStrings[binaryName]!!
+        else -> binaryName.replace('$', '.')
+    }
+
+
+private
+val collectionTypeStrings =
+    mapOf(
+        "java.util.Iterable" to "kotlin.collections.Iterable",
+        "java.util.Iterator" to "kotlin.collections.Iterator",
+        "java.util.Collection" to "kotlin.collections.Collection",
+        "java.util.List" to "kotlin.collections.List",
+        "java.util.ArrayList" to "kotlin.collections.ArrayList",
+        "java.util.Set" to "kotlin.collections.Set",
+        "java.util.HashSet" to "kotlin.collections.HashSet",
+        "java.util.LinkedHashSet" to "kotlin.collections.LinkedHashSet",
+        "java.util.Map" to "kotlin.collections.Map",
+        "java.util.HashMap" to "kotlin.collections.HashMap",
+        "java.util.LinkedHashMap" to "kotlin.collections.LinkedHashMap")
```

Comment:
Does Kotlin have this mapping somewhere in their compiler that you could use instead?
It may not be part of their public API.
Also, what about the the other collection types like:
`NavigableMap` `SortedSet`, `TreeSet`, ect...?

Reply:
`NavigableMap`, `SortedSet`, `TreeSet` are not part of `kotlin.collections`, when you use them in Kotlin you the original ones from `java.util.*`

We should at least make sure we conform to https://kotlinlang.org/docs/reference/java-interop.html#mapped-types